hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/LocalhostServingController.java

```diff
@@ -0,0 +1,398 @@
+      // Update the object in the database
+      Serving dbServing = servingFacade.updateDbObject(serving, project);
+      if (status == ServingStatusEnum.RUNNING || status == ServingStatusEnum.UPDATING) {
+        if (!oldDbServing.getName().equals(dbServing.getName()) ||
+            !oldDbServing.getArtifactPath().equals(dbServing.getArtifactPath()) ||
+            oldDbServing.isBatchingEnabled() != dbServing.isBatchingEnabled() ||
+            oldDbServing.getVersion() > dbServing.getVersion()) {
+          // To update the name and/or the artifact path we need to restart the server and/or the version as been
+          // reduced. We need to restart the server
+          restartServingInstance(project, user, oldDbServing, dbServing);
+        } else {
+          // To update the version call the script and download the new version in the directory
+          // the server polls for new versions and it will pick it up.
+          if(serving.getServingType() == ServingType.TENSORFLOW){
+            tfServingController.updateModelVersion(project, user, dbServing);
+          } else {
+            //If we do not need to update model version there is nothing left to do and we can release the lock
+            servingFacade.releaseLock(project, serving.getId());
+          }
+        }
+      } else {
+        // The instance is not running, nothing else to do. Just release the lock.
+        servingFacade.releaseLock(project, serving.getId());
+      }
+    }
+  }
+  
+  /**
+   * Check if there is already a serving with the same name as a new/updated serving in the project
+   *
+   * @param project the project to query
+   * @param servingWrapper the serving to compare with the existing servings
+   * @throws ServingException if a duplicate was found in the database
+   */
+  @Override
+  public void checkDuplicates(Project project, ServingWrapper servingWrapper) throws ServingException {
+    Serving serving = servingFacade.findByProjectAndName(project,
+        servingWrapper.getServing().getName());
+    if (serving != null && !serving.getId().equals(servingWrapper.getServing().getId())) {
+      // There is already an entry for this project
+      throw new ServingException(RESTCodes.ServingErrorCode.DUPLICATEDENTRY, Level.FINE);
+    }
+  }
+  
+  /**
+   * Validates user input before creating or updating a serving. This method contains the common input validation
+   * between different serving types and then delegates to serving-type-specific controllers to do validation specific
+   * to the serving type.
+   *
+   * @param servingWrapper contains the user-data to validate
+   * @param project the project where the serving resides
+   * @throws ServingException
+   */
+  @Override
+  public void validateUserInput(ServingWrapper servingWrapper, Project project) throws ServingException {
+    // Check that the modelName is present
+    if (Strings.isNullOrEmpty(servingWrapper.getServing().getName())) {
```

Comment:
Any way we can avoid code duplication between the LocalhostServingController and the K8sServingController?

Reply:
Done, moved common stuff to a Util Bean